File: Simulation1.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
from matplotlib import style
import os
import re

style.use("ggplot")

SIZE = 20
EPISODES = 10000
MOVE_PENALTY = 1
MOUNTAIN_REWARD = 300  
EPSILON = 0.9
EPSILON_DECAY = 0.9998
SHOW_EVERY = 500
LEARNING_RATE = 0.1
DISCOUNT = 0.95

TARGET = (SIZE - 1, SIZE // 2)

if 'q_table' not in globals():
    q_table = {}
    for x1 in range(-SIZE + 1, SIZE):
        for y1 in range(-SIZE + 1, SIZE):
            q_table[(x1, y1)] = [np.random.uniform(-5, 0) for i in range(4)]

episode_rewards = []
CHARACTER_1 = 1
PASTEL = {1: (0, 255, 0)}

DIRECTORY = "Data S1"
if not os.path.exists(DIRECTORY):
    os.makedirs(DIRECTORY)

class AGENT:
    def __init__(self):
        self.x = 0
        self.y = SIZE - 1

    def action(self, choice):
        if choice == 0:
            self.move(x=1, y=0)
        elif choice == 1:
            self.move(x=-1, y=0)
        elif choice == 2:
            self.move(x=0, y=1)
        elif choice == 3:
            self.move(x=0, y=-1)

    def move(self, x=False, y=False):
        self.x = np.clip(self.x + x, 0, SIZE - 1)
        self.y = np.clip(self.y + y, 0, SIZE - 1)

def mountain(size):
    mountain = np.zeros((size, size))
    peak_height = size // 4
    base_width = size // 4
    base_start_col = size - base_width
    for i in range(peak_height):
        start = base_start_col + (peak_height - i - 1)
        end = size - (peak_height - i - 1)
        mountain[i, start:end] = 1
    return mountain

mountain_render = mountain(SIZE)

def mountain_collision(character_x, character_y, mountain_grid):
    return mountain_grid[character_y, character_x] > 0

def check_if_reached_target(character_x, character_y):
    return mountain_collision(character_x, character_y, mountain_render)

def next_run_number(directory):
    max_num = 0
    for filename in os.listdir(directory):
        match = re.match(r"Episode_rewards_(\d+).jpg", filename)
        if match:
            num = int(match.group(1))
            if num > max_num:
                max_num = num
    return max_num + 1

run_number = next_run_number(DIRECTORY)
filename = f"{DIRECTORY}/Episode_rewards_{run_number}.jpg"

for episode in range(EPISODES):
    character = AGENT()
    
    if episode % SHOW_EVERY == 0:
        print(f"on #{episode}, epsilon is {EPSILON}")
        show = True
    else:
        show = False
    
    episode_reward = 0
    
    for i in range(200):
        obs = (character.x - TARGET[0], character.y - TARGET[1])
        if np.random.random() > EPSILON:
            action = np.argmax(q_table[obs])
        else:
            action = np.random.randint(0, 4)
        character.action(action)

        reward = -MOVE_PENALTY
        if mountain_collision(character.x, character.y, mountain_render):
            reward = MOUNTAIN_REWARD

        new_obs = (character.x - TARGET[0], character.y - TARGET[1])
        max_future_q = np.max(q_table.get(new_obs, np.zeros(4)))
        current_q = q_table[obs][action]
        if reward == MOUNTAIN_REWARD:
            new_q = MOUNTAIN_REWARD
        else:
            new_q = (1 - LEARNING_RATE) * current_q + LEARNING_RATE * (reward + DISCOUNT * max_future_q)
        q_table[obs][action] = new_q
        
        episode_reward += reward

        if check_if_reached_target(character.x, character.y):
            break
        
        if show:
            env = np.full((SIZE, SIZE, 3), 128, dtype=np.uint8)
            mountain_env = env.copy()
            mountain_env[mountain_render > 0] = (0, 0, 255)
            mountain_env = cv2.resize(mountain_env, (300,300), interpolation=cv2.INTER_NEAREST)
            character_position = (int(character.x * 300 / SIZE), int(character.y * 300 / SIZE))
            cv2.circle(mountain_env, character_position, 15, PASTEL[CHARACTER_1], -1)
            cv2.imshow("Mountain Env", mountain_env)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

    episode_rewards.append(episode_reward)
    EPSILON *= EPSILON_DECAY

reward_array = np.array(episode_rewards)
smoothed_rewards = np.convolve(reward_array, np.ones((SHOW_EVERY,))/SHOW_EVERY, mode='valid')

plt.plot(np.arange(len(smoothed_rewards)), smoothed_rewards)
plt.ylim(0, max(smoothed_rewards) + 20)
plt.xlabel('Episode #')
plt.ylabel('Reward')
plt.savefig(filename)
plt.show()
cv2.destroyAllWindows()

File: Simulation1.js
document.addEventListener("DOMContentLoaded", function () {
    const canvas = document.getElementById('envCanvas');
    const ctx = canvas.getContext('2d');
    const infoBox = document.getElementById('info');
    const SIZE = 20;
    const MOVE_PENALTY = 1;
    const MOUNTAIN_REWARD = 300;
    const EPSILON_DECAY = 0.9998;
    const LEARNING_RATE = 0.1;
    const DISCOUNT = 0.95;
    const SCALE = canvas.width / SIZE;
    let epsilon = 0.9;
    let qTable = {};
    let agent = new Agent();
    let resetCount = 0;
    let mountainTouches = 0;
    let simulationActive = true;

    function Agent() {
        this.x = 0;
        this.y = SIZE - 1;
    }

    Agent.prototype.action = function(choice) {
        if (choice === 0) this.x = Math.max(this.x - 1, 0);
        else if (choice === 1) this.x = Math.min(this.x + 1, SIZE - 1);
        else if (choice === 2) this.y = Math.max(this.y - 1, 0);
        else if (choice === 3) this.y = Math.min(this.y + 1, SIZE - 1);
    };

    function createMountain() {
        let mountain = Array.from({length: SIZE}, () => Array(SIZE).fill(0));
        const peakHeight = Math.floor(SIZE / 4);
        const baseWidth = Math.floor(SIZE / 4);
        const baseStartCol = SIZE - baseWidth;
        for (let i = 0; i < peakHeight; i++) {
            for (let j = baseStartCol + (peakHeight - i - 1); j < SIZE - (peakHeight - i - 1); j++) {
                mountain[i][j] = 1;
            }
        }
        return mountain;
    }

    let mountainRender = createMountain();

    function draw() {
        ctx.fillStyle = '#808080';
        ctx.fillRect(0, 0, canvas.width, canvas.height);
        for (let y = 0; y < SIZE; y++) {
            for (let x = 0; x < SIZE; x++) {
                ctx.fillStyle = mountainRender[y][x] === 1 ? 'red' : '#808080';
                ctx.fillRect(x * SCALE, y * SCALE, SCALE, SCALE);
            }
        }
        ctx.beginPath();
        ctx.arc(agent.x * SCALE + SCALE / 2, agent.y * SCALE + SCALE / 2, SCALE / 2, 0, 2 * Math.PI);
        ctx.fillStyle = 'green';
        ctx.fill();
        infoBox.textContent = `Resets: ${resetCount}, Mountain Touches: ${mountainTouches}, Agent Position: (${agent.x}, ${agent.y}), Epsilon: ${epsilon.toFixed(4)}`;
    }

    function simulate() {
        if (!simulationActive) return;
        let action = Math.floor(Math.random() * 4);
        agent.action(action);
        let collision = mountainRender[agent.y][agent.x] === 1;
        if (collision) {
            mountainTouches++;
            simulationActive = false;
        }
        let reward = collision ? MOUNTAIN_REWARD : -MOVE_PENALTY;
        let obs = `${agent.x},${agent.y}`;
        if (!qTable[obs]) qTable[obs] = Array(4).fill().map(() => Math.random() * -5);
        let maxFutureQ = Math.max(...qTable[obs]);
        let currentQ = qTable[obs][action];
        qTable[obs][action] = (1 - LEARNING_RATE) * currentQ + LEARNING_RATE * (reward + DISCOUNT * maxFutureQ);
        epsilon *= EPSILON_DECAY;
        draw();
        setTimeout(simulate, 100);
    }

    window.resetSimulation = function() {
        simulationActive = true;
        resetCount++;
        agent = new Agent();
        simulate();
    };

    window.clearMemory = function() {
        epsilon = 0.9;
        qTable = {};
        mountainTouches = 0;
        resetCount = 0;
        simulationActive = true;
        simulate();
    };

    simulate();
});